Add Vulkan Tensor.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44017 Add Vulkan op Add.
* #46481 Add Vulkan tensor copy.
* #44016 Add Vulkan Tensor factory.
* **#44015 Add Vulkan Tensor.**

Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Resource.h:
```diff
@@ -27,22 +28,49 @@ struct Resource final {
+      template<typename Type, Flags access>
+      using Pointer = std::add_pointer_t<
+          std::conditional_t<
+              0u != (access & Write),
```
Could it be juts: `access & Write`?

AshkanAliabadi replied:
That's what I thought originally as well but unfortunately the compiler wasn't happy with that and kept giving me cryptic errors which took quite a while to root cause. :/  The first argument of std::conditional is a boolean and I am not quite sure why an implicit integer ((access & Write) returns an integer) to boolean conversion could not have been performed automatically in that context like everywhere else, so had to make this explicitly a boolean with a comparison against zero.

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Resource.h:
```diff
@@ -354,21 +394,29 @@ inline Resource::Fence::operator bool() const {
+        Block,
+        Memory::Access::Write>();
+
+    *memory.get() = block;
```
Am I right that semantic of type Block here is  something that has sizeof and and can be assigned to memory pointer. Will it copy here all the block data to allocated memory?

AshkanAliabadi replied:
Yes, almost.  Invocation of sizeof on any data type or variable in C++ is valid.  As for the second part of your statement, if Block has an assignment operator defined, whether implicitly by the compiler such as the cases where Block is a simple POD as is all of our use cases, or explicitly defined by the user if Block has more complex semantics, then this code will compile.  The reason behind this implementation is two fold:  The primary motivation is that memcpy is a function call unless treated specifically by the compiler as an intrinsic considering how common place it is, and calling a complex function to copy a few bytes is overkill.  The second, in this particular case less important, reason is that memcpy is only technically a valid operation on PODs, otherwise the assignment operator should be called for proper treatment of non primitive types.

In the future we should move to push constants where possible.  That will inline parameter passing (for small blocks) inside the command buffer and should be much faster compared to all the overhead of allocation a uniform buffer (even from a pool as do), creation of vkBuffer, mapping and unmapping of that memory, all for a few bytes of data.

IvanKobzarev commented on aten/src/ATen/native/vulkan/ops/Tensor.cpp:
```diff
@@ -0,0 +1,1230 @@
+  }();
+
+  const auto memory = [context]() -> api::Resource::Memory::Descriptor {
+    if (requires_staging(context)) {
```
What is the semantic of requires_staging here,
From the first impression I was thinking that it means 'staging' buffer that should  be HOST_VISIBLE, but here I see that this is inverted and `if !requires_staging => HOST_VISIBLE` 
Is it right and I am missing something ?

AshkanAliabadi replied:
We will need staging buffers for discrete GPUs only.   Under those circumstances, we will have a staging buffer in system memory, a buffer in GPU memory, and an image, again, in GPU memory.  If on UMA, we can get away with only a buffer and an image, with the buffer accessible both from CPU and GPU.

What this line says is that if we require a staging buffer, i.e. we are on a discrete system, then create the buffer copy in GPU memory only (this function, allocate_buffer, only deals with the allocation of GPU-visible buffer not CPU-visible staging area), but if we do not require a staging area, meaning that we are on UMA, then make sure this GPU-visible memory / buffer is also CPU-visible, which is handled in the (implicit) 'else' part of this if statement.

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/ops/Tensor.h:
```diff
@@ -0,0 +1,573 @@
+    Future
+  */
+
+  template<typename Type, Access::Flags kAccess>
```
Nit: Usually I saw naming with k prefix for constexpr's. Am I missing something about the meaning of that k prefix? :)

AshkanAliabadi replied:
No, you are right.  I will revert.  :)

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan Tensor."


Differential Revision: [D23820827](https://our.internmc.facebook.com/intern/diff/D23820827)

[ghstack-poisoned]
